src/traces/treemap/draw_descendants.js

```diff
@@ -65,16 +65,25 @@ module.exports = function drawDescendants(gd, cd, entry, slices, opts) {
 
     var sliceData = allData.descendants();
 
+    var minVisibleDepth = Infinity;
+    var maxVisibleDepth = -Infinity;
+
     slices = slices.data(sliceData, function(pt) {
-        // hide slices that won't show up on graph
-        if(pt.depth >= trace._maxDepth) {
+        var depth = pt.depth;
+        if(depth >= trace._maxDepth) {
+            // hide slices that won't show up on graph
             pt.x0 = pt.x1 = (pt.x0 + pt.x1) / 2;
             pt.y0 = pt.y1 = (pt.y0 + pt.y1) / 2;
+        } else {
+            minVisibleDepth = Math.min(minVisibleDepth, depth);
+            maxVisibleDepth = Math.max(maxVisibleDepth, depth);
         }
```

Comment:
Please bring this out of the key function. 

You can have something like:

```js
 var minVisibleDepth = Infinity;
 var maxVisibleDepth = -Infinity;
sliceData.forEach(function(pt) {
        var depth = pt.depth;
        if(depth >= trace._maxDepth) {
            // hide slices that won't show up on graph
            pt.x0 = pt.x1 = (pt.x0 + pt.x1) / 2;
            pt.y0 = pt.y1 = (pt.y0 + pt.y1) / 2;
        } else {
            minVisibleDepth = Math.min(minVisibleDepth, depth);
            maxVisibleDepth = Math.max(maxVisibleDepth, depth);
        }
})

slices = slices.data(sliceData, function(pt) { return helpers.getPtId(pt); });
// or even ;)
slices = slices.data(sliceData, helpers.getPtId)
```

Reply:
Thank you!
Fixed in https://github.com/plotly/plotly.js/pull/4242/commits/280a0a1baf7b7c9e8c4965599ed96402a38fa746. Also revised draw_ancestors and sunburst functions.

Reply:
Awesome!!